### PR TITLE
[Feature]Virtual Pipeline Parallel

### PR DESCRIPTION
**Pull request by RookieCoder-Camera**
What this PR does / why we need it?
This PR adds Virtual Pipeline Parallelism (VPP) support to vllm-ascend on
top of vLLM v0.22.1, implemented as pure-Python monkey-patches with zero
modification to the upstream vLLM core. It is opt-in and fully backward
compatible (default vp_size=1 leaves all code paths unchanged).

VPP splits one logical forward into vp_size virtual stages distributed across
PP ranks using a V-shaped fold-back topology (even virtual stages sweep
forward rank 0 → N-1, odd stages sweep backward rank N-1 → 0). At the fold
point the same physical rank continues to the next virtual stage with no
inter-rank communication. This overlaps communication with computation across
virtual stages and shrinks the pipeline bubble versus plain PP — measured to
outperform Context-Parallel-Pipeline (CPP) on DeepSeek-V3.1
(see How was this patch tested?).

User-facing config (all opt-in):

--additional-config '{"virtual_pipeline_parallel_size": 2,
                      "vpp_layer_ranges": [[..], [..]]}'   // optional manual override
ENABLE_VPP=true                                         // env: VPP-aware output-rank routing
Key implementation pieces:

┌──────────────┬──────────────────────────────────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Area │ File(s) │ What it does │
├──────────────┼──────────────────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Topology │ distributed/vpp_utils.py (new) │ V-fold layer assignment (get_vpp_indices), per-stage comm plan (get_vpp_comm_info), fold-point / │
│ math │ │ last-stage predicates │
├────────────────┼────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Global state │ distributed/parallel_state.py │ vp_rank/vp_size globals + the _VPP_RUNTIME_ACTIVE construction-vs-runtime latch │
├────────────────┼────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Config │ ascend_config.py, envs.py │ virtual_pipeline_parallel_size + vpp_layer_ranges validation; ENABLE_VPP env switch │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ EngineCore │ patch/platform/patch_vpp_core.py │ Replaces step_with_batch_queue with a fold-back dispatch loop that re-issues execute_model on │
│ │ │ VppContinuationOutput │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Scheduler │ patch/platform/patch_scheduler.py, │ Stamps monotonic batch_id on SchedulerOutput; injects pickle-safe VppContinuationOutput │
│ │ patch_outputs.py │ │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Executor │ patch/platform/patch_multiproc_executor.py │ Routes ModelRunnerOutput to the correct global rank per V-fold topology │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Worker │ worker/worker.py, worker/model_runner_v1.py │ execute_model dispatch; the resumable _run_vpp_stage loop driving one virtual stage per RPC, │
│ │ │ keyed by batch_id via _vpp_contexts │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Distributed │ patch/worker/patch_distributed.py │ Phase-latched is_first_rank/is_last_rank (so upstream init places embed/norm/lm_head on │
│ │ │ the correct physical rank) + fully-async send_tensor_dict │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Layer build │ patch/worker/patch_vpp_make_layers.py │ One centralized patch on the shared make_layers → all models get V-fold layer materialization │
│ │ │ for free │
├────────────────┼─────────────────────────────────────────────────┼────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Per-stage │ ascend_forward_context.py, │ Scope the MoE layer list per virtual stage; snapshot sampling state (embed stage and sampling │
│ support │ worker/npu_input_batch.py │ stage land in different RPCs) │
└────────────────┴─────────────────────────────────────────────────┴────────────────────────────────────────────────────────────────────────────────────────────────┘

Design highlights worth noting for review:

Resumable execution across RPCs. A single logical forward spans multiple
execute_model calls. The worker yields VppContinuationOutput between
virtual stages; EngineCore.step_with_batch_queue re-issues the same
SchedulerOutput (matched by batch_id) until the last stage produces real
logits. Per-batch progress lives in _vpp_contexts: dict[batch_id, ctx].
Construction-vs-runtime latch. Upstream model init places
norm/lm_head on is_last_rank, which at construction time (vp_rank=0)
would be wrong for every rank. _VPP_RUNTIME_ACTIVE makes
is_first_rank/is_last_rank return the static fold-back topology during
init and the dynamic current-virtual-stage answer during forward.
This sinks topology correctness into the rank predicates themselves, so no
per-model init patch is needed — every model that builds its decoder
via the shared make_layers is VPP-correct.
Fold-point deadlock avoidance. The matching recv of a fold-point
send lands in a later RPC iteration, which would deadlock on a single
serialized HCCL stream. Mitigated by (a) two PP comm groups for the
forward/backward sweeps, (b) fully async isend-based send_tensor_dict
with a FIFO send buffer holding source tensors until HCCL completion
Does this PR introduce any user-facing change?

Yes — additive and opt-in only; no change when vp_size == 1:

New additional_config keys: virtual_pipeline_parallel_size (int, default
1; >1 requires pipeline_parallel_size > 1) and optional
vpp_layer_ranges (manual per-rank, per-stage [start, end) override,
validated for full, non-overlapping coverage).
New environment variable ENABLE_VPP (default false), documented in
vllm_ascend/envs.py. Required to be true (or DYNAMIC_EPLB=true) whenever
virtual_pipeline_parallel_size > 1, so the multiproc executor routes outputs
to the correct PP rank.
No public API/type removal or rename; no change to default behavior.
How was this patch tested?

Functional bring-up on a 16-card Ascend NPU node, DeepSeek-V3.1-Terminus
-w8a8-QuaRot-lfs, TP=8 × PP=2 × vp_size=2:

export ENABLE_VPP=true
vllm serve <DeepSeek-V3.1-Terminus-w8a8-QuaRot-lfs>
--tensor-parallel-size 8 --pipeline-parallel-size 2
--enable-expert-parallel --max-num-seqs 16
--max-model-len 527360 --max-num-batched-tokens 16384
--gpu-memory-utilization 0.88 --enable-chunked-prefill
--enforce-eager
--additional-config '{"virtual_pipeline_parallel_size": 2}'

Verified from the server log:

Correct uneven layer split: VPP enabled (uneven): vp_size=2, pp_size=2, num_layers=61, 1 chunks get 16 layers, 3 chunks get 15 layers.
Construction-phase latch fires correctly on all 16 ranks (is_first_rank
True on PP0 ranks, False on PP1 ranks) → embed/norm/lm_head land on the
right physical ranks, no PPMissingLayer unpack errors.
Application startup complete; the engine serves requests end-to-end across
the multi-RPC virtual-stage loop.
Performance vs CPP: VPP (TP8 × PP2 × vp2) outperforms Context-Parallel-
Pipeline (TP8 × CPP2) on the same model under our offline throughput
workload. [TODO: insert your measured numbers — e.g. prefill/decode tokens/s
for VPP vs CPP, same concurrency/input length.]

Test gaps / plan. End-to-end VPP correctness is inherently multi-process
and NPU-dependent, which makes a pure UT hard. As a follow-up I plan to add UTs
for the pure topology functions that are hardware-independent:
get_vpp_indices (even/uneven coverage), get_vpp_comm_info
(fold-point recv/send direction), and is_vpp_last_stage/is_vpp_fold_point
for both even and odd vp_size. Happy to include them in this PR if reviewers
prefer.

New additional_config keys: virtual_pipeline_parallel_size (int, default
1; >1 requires pipeline_parallel_size > 1) and optional
vpp_layer_ranges (manual per-rank, per-stage [start, end) override,
validated for full, non-overlapping coverage).
New environment variable ENABLE_VPP (default false), documented in
vllm_ascend/envs.py. Required to be true (or DYNAMIC_EPLB=true) whenever
virtual_pipeline_parallel_size > 1, so the multiproc executor routes outputs
to the correct PP rank.
No public API/type removal or rename; no change to default behavior.
How was this patch tested?

Functional bring-up on a 16-card Ascend NPU node, DeepSeek-V3.1-Terminus
-w8a8-QuaRot-lfs, TP=8 × PP=2 × vp_size=2:

export ENABLE_VPP=true
vllm serve <DeepSeek-V3.1-Terminus-w8a8-QuaRot-lfs>
--tensor-parallel-size 8 --pipeline-parallel-size 2
--enable-expert-parallel --max-num-seqs 16
--max-model-len 527360 --max-num-batched-tokens 16384
--gpu-memory-utilization 0.88 --enable-chunked-prefill
--enforce-eager
--additional-config '{"virtual_pipeline_parallel_size": 2}'

Verified from the server log:

Correct uneven layer split: VPP enabled (uneven): vp_size=2, pp_size=2, num_layers=61, 1 chunks get 16 layers, 3 chunks get 15 layers.
Construction-phase latch fires correctly on all 16 ranks (is_first_rank
True on PP0 ranks, False on PP1 ranks) → embed/norm/lm_head land on the
right physical ranks, no PPMissingLayer unpack errors.
Application startup complete; the engine serves requests end-to-end across
the multi-RPC virtual-stage loop.
Performance vs CPP: VPP (TP8 × PP2 × vp2) outperforms Context-Parallel-
Pipeline (TP8 × CPP2) on the same model under our offline throughput
workload. [TODO: insert your measured numbers — e.g. prefill/decode tokens/s
for VPP vs CPP, same concurrency/input length.]

Test gaps / plan. End-to-end VPP correctness is inherently multi-process
and NPU-dependent, which makes a pure UT hard. As a follow-up I plan to add UTs
for the pure topology functions that are hardware-independent:
get_vpp_indices (even/uneven coverage), get_vpp_comm_info
(fold-point recv/send direction), and is_vpp_last_stage/is_vpp_fold_point
for both even and odd vp_size. Happy to include them in this PR if reviewers
prefer.

CoAuthor:
[wangxiaochao6@hisilicon.com](mailto:wangxiaochao6@hisilicon.com)
[gaojingchun1@huawei.com](mailto:gaojingchun1@huawei.com)
[luojintao5@huawei.com](mailto:luojintao5@huawei.com)
[lixushi@huawei.com](mailto:lixushi@huawei.com)
[liufeng248@huawei.com](mailto:liufeng248@huawei.com)
[shiyan67@huawei.com](mailto:shiyan67@huawei.com)

vLLM version: v0.24.0
vLLM main: https://github.com/vllm-project/vllm/commit/85c09e9885e346ea1612da30ebff5a75f67d2350

- vLLM main: https://github.com/vllm-project/vllm/commit/e6bfe03ad73a3330cb427885aa90d97a12e1c704

### CODE REVIEW

**Comment by zhou9402** [comment]
Hi @RookieCoder-Camera , is there any data to show the benefit of VPP?
